### ISSUE TITLE
UTF-8 filenames break GodMode9

### ISSUE THREAD

**Issue by NWPlayer123** [d0k3/GodMode9#368]
I was trying to extract an NDS rom (Nintendogs Dachshund and Friends), which has a file called ＵＳＡまだっす.txt in the root, it gets confused and refuses to copy the rest to my SD

**Comment by d0k3** [comment]
I'll look into. Guess that's a DS specific problem, it was already solved for 3DS files.

**Comment by urherenow** [comment]
Not that I know what I'm doing, but does this commit on fastboot3ds help at all?
https://github.com/derrekr/fastboot3DS/commit/c23b0e1cc71f1d28e3fb7cbf01197c4cdcb05e7d

**Comment by d0k3** [comment]
@urherenow - that's only about the FatFS lib, and GM9 is already up to date with that. That can't be the problem here.

@NWPlayer123 - I tried it, and, yup, it's weird. GM9 should be perfectly capable of handling UTF-8 in copy operations. Looking into it further...

**Comment by d0k3** [comment]
Okay, after some further looking into... GM9 is perfectly capable of handling standard UTF-8 filenames and even UTF-16 - try the 3DS Fire Emblem Fates games for example (actually, try one of them, just to make sure theres no problem on your end).

Now, as for that one... what's that at the beginning? Emoji letters? Unsure if FAT can even handle that. Can you name a tool that would actually extract that file from the DS rom properly?

**Comment by NWPlayer123** [comment]
like I said, it's called ＵＳＡまだっす.txt, unicode "USA", tinke happily extracts it

**Comment by d0k3** [comment]
Okay, keeping you up to date on my progress. This is not UTF-8, it's Shift-JIS, which is bad. A proper copy to FAT would mean proper UFT-8 <-> Shift-JIS conversion, and that's not exactly trivial. Maybe I'll look into just a workaround, seeing how this is not exactly common.

**Comment by pthfdr-42** [comment]
I think that a better fallback is to screw all Unicode and display raw bytes with CP437.

**Comment by ParzivalWolfram** [comment]
Yeah, Shift-JIS is a pain. I'd recommend just "converting" by either filename truncation or just allowing it to become mojibake.

**Comment by pthfdr-42** [comment]
Seriously why they are still using an encoding other than UTF-8?
In China, the same problem with GB2312 and GBK.
NIH syndrome at its finest.

**Comment by knight-ryu12** [comment]
JP here, because Windows.

**Comment by pthfdr-42** [comment]
I have to say that Windows seems to avoid UTF encodings for a reason. Hmm. (I use Debian Sid and use the terminal for nearly everything)
`set RANT=1`
I once copied some audio files from my dormmates Windows PC. When I plug in the USB disk, both the filenames and the metadata are messed up. (`urxvt` supports Unicode) Tried to archive the files but the filenames are still corrupted when extracted. Solution? I set up an FTP server on his computer and "downloaded" the files over the router. The metadata is still corrupted but I do not rely on them very much.
`unset RANT`

**Comment by ParzivalWolfram** [comment]
Windows usually has problems with other languages due to avoiding Unicode. Locales are a pain in the ass...
